ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ACSReader.scala

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.appendonlydao.events
+
+import akka.NotUsed
+import akka.stream.{BoundedSourceQueue, Materializer, QueueOfferResult}
+import akka.stream.scaladsl.Source
+import com.daml.dec.DirectExecutionContext
+import com.daml.ledger.api.v1.active_contracts_service.GetActiveContractsResponse
+import com.daml.ledger.offset.Offset
+import com.daml.lf.data.Ref
+import com.daml.logging.LoggingContext
+
+import scala.collection.mutable
+import scala.concurrent.Future
+
+trait ACSReader {
+  def acsStream(
+      filter: FilterRelation,
+      activeAt: (Offset, Long),
+      verbose: Boolean,
+  )(implicit loggingContext: LoggingContext): Source[GetActiveContractsResponse, NotUsed]
+}
+
+class FilterTableACSReader extends ACSReader {
+
+  override def acsStream(
+      filter: FilterRelation,
+      activeAt: (Offset, Long),
+      verbose: Boolean,
+  )(implicit loggingContext: LoggingContext): Source[GetActiveContractsResponse, NotUsed] =
+    throw new NotImplementedError() // FIXME will be implemented in a later increment
+
+}
+
+private[events] object FilterTableACSReader {
+
+  case class Filter(party: Party, templateId: Option[Ref.Identifier])
+
+  case class QueryTask(fromExclusiveEventSeqId: Long, filter: Filter)
+
+  object QueryTask {
+    implicit val ordering: Ordering[QueryTask] =
+      Ordering.by[QueryTask, Long](_.fromExclusiveEventSeqId)
+  }
+
+  /** This Source implementation solves the following problem:
+    *  - let us have n TASKs, which are ordered
+    *  - let us have an work defined over these tasks, which gives us a RESULT and a possible continuation of the TASK
+    *  - let us have configurable parallelism to work on these TASKs
+    *  This implementation ensures that all the time the "smallest" available TASK will be picked for work.
```

Comment:
I think "smallest" doesn't need to be quoted. You have introduced the concept of ordering in the first bullet point.
What happens with a task that was smallest and finished with continuation. Does it go back to the front of the queue?

Reply:
> I think "smallest" doesn't need to be quoted. 
yes, good point, will do

> What happens with a task that was smallest and finished with continuation. Does it go back to the front of the queue?
If a task finished with a continuation it will be put back to the queue. The queue is a priority queue, so it will be put in the right place based on the ordering. As picking the next task, the smallest task will be chosen.